File: content/5-memory/2-history-compaction/openai/1-sliding_window.py
from openai import OpenAI

LLM_MODEL = "gpt-5.4"
client = OpenAI()

# sliding window: keep last N messages, discard everything older
# cheapest — zero latency, but complete context loss beyond the window
WINDOW = 10
messages: list = []

def chat(message: str) -> str:
    messages.append({"role": "user", "content": message})
    # only send the last WINDOW messages
    window = messages[-WINDOW:]
    response = client.responses.create(model=LLM_MODEL, input=window)
    messages.append({"role": "assistant", "content": response.output_text})
    return response.output_text

print(chat("What is the capital of France?"))
print(chat("What is its population?"))

File: content/5-memory/2-history-compaction/aisdk/1-sliding_window.ts
import { generateText, type ModelMessage } from "ai";
import { openai } from "@ai-sdk/openai";

const LLM_MODEL = "gpt-5.4";

// sliding window: keep last N messages, discard everything older
// cheapest — zero latency, but complete context loss beyond the window
const WINDOW = 10;
const messages: ModelMessage[] = [];

async function chat(message: string): Promise<string> {
  messages.push({ role: "user", content: message });
  const window = messages.slice(-WINDOW);
  const result = await generateText({
    model: openai(LLM_MODEL),
    messages: window,
  });
  messages.push({ role: "assistant", content: result.text });
  return result.text;
}

console.log(await chat("What is the capital of France?"));
console.log(await chat("What is its population?"));
